core/trino-main/src/main/java/io/trino/connector/WorkerDynamicCatalogManager.java

```diff
@@ -100,14 +105,19 @@ public void ensureCatalogsLoaded(Session session, List<CatalogProperties> expect
                 return;
             }
 
-            for (CatalogProperties catalog : getMissingCatalogs(expectedCatalogs)) {
-                checkArgument(!catalog.getCatalogHandle().equals(GlobalSystemConnector.CATALOG_HANDLE), "Global system catalog not registered");
-                catalogs.computeIfAbsent(catalog.getCatalogHandle(), ignore -> {
-                    CatalogConnector newCatalog = catalogFactory.createCatalog(catalog);
-                    log.debug("Added catalog: " + catalog.getCatalogHandle());
-                    return newCatalog;
-                });
-            }
+            CompletableFuture.allOf(
+                getMissingCatalogs(expectedCatalogs).stream()
+                        .map(catalog ->
+                                CompletableFuture.runAsync(() -> {
+                                    checkArgument(!catalog.getCatalogHandle().equals(GlobalSystemConnector.CATALOG_HANDLE), "Global system catalog not registered");
```

Comment:
Assign, `getMissingCatalogs(expectedCatalogs)` to a varaible outside of the `allOf` call.  Also verify that none of the mising catalogs are global system catalogs outside of `allOf`.